lexer/src/main/kotlin/lexer/Lexer.kt

```diff
@@ -0,0 +1,318 @@
+    private var currentPosition = FileOffset(charIndex = -1, row = 0, col = -1)
+
+    fun readTokenStream(): Sequence<Token> = sequence {
+        // Read very first character
+        readNext()
+        skipBlanks()
+        while (!end) {
+            val token = readToken()
+            if (token != null) {
+                yield(token)
+            }
+            skipBlanks()
+        }
+    }
+
+    private fun readToken(): Token? = when {
+        current.isDigit() -> readNumber()
+        current.isLetter() -> readName()
+        else -> when (current) {
+            '-' -> readNumber(isNegative = true)
+            '"' -> readString()
+            '/' -> skipComment()
+            '.' -> readDot()
+            '^' -> readName(caretPassed = true)
+            '{' -> readStaticToken(Tag.OpenBrace)
+            '}' -> readStaticToken(Tag.CloseBrace)
+            '[' -> readStaticToken(Tag.OpenBracket)
+            ']' -> readStaticToken(Tag.CloseBracket)
+            '(' -> readStaticToken(Tag.OpenParenthesis)
+            ')' -> readStaticToken(Tag.CloseParenthesis)
+            ',' -> readStaticToken(Tag.Comma)
+            ':' -> readStaticToken(Tag.Colon)
+            '*' -> readStaticToken(Tag.Asterisk)
+            '@' -> readStaticToken(Tag.AtSign)
+            '<' -> readStaticToken(Tag.LessThanSign)
+            '>' -> readStaticToken(Tag.GreaterThanSign)
+            '?' -> readStaticToken(Tag.QuestionMark)
+            else -> {
+                val start = currentPosition
+                readNext() // Skip unrecognized character
+                val errorLocation = locationFromStart(start)
+                diagnostics.reportError("Unrecognized character: $current", errorLocation)
+                null
+            }
+        }
+    }
+
+    private fun readStaticToken(tag: Tag): StaticToken {
+        val start = currentPosition
+        readNext()
+        return StaticToken(locationFromStart(start), tag)
+    }
+
+    private fun readNumber(isNegative: Boolean = false): NumberToken {
+        val start = currentPosition
+
+        var hasDecimalPoint = false
+        val numberAsString = buildString {
+            fun readDigits(): Boolean {
+                val previousLength = length
+                while (!end && current.isDigit()) {
+                    append(current)
+                    readNext()
+                }
+                return length > previousLength
+            }
+
+            if (isNegative) {
+                assert(current == '-')
+                append(current)
+                readNext() // Ignore minus for number
+            }
+            val hasWholeDigits = readDigits()
+            if (!hasWholeDigits) {
+                diagnostics.reportError("Number is missing whole part (0.5 is valid, .5 is not)", locationFromStart(start))
+                append('0')
+            }
+
+            if (current == '.') {
+                reader.mark(1)
+                assert(reader.markSupported())
```

Comment:
mark would throw an IOException anyways if it's not supported
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/Reader.html#mark(int)

Reply:
Removed the `assert`